be/src/runtime/stream_load/transaction_mgr.cpp

```diff
@@ -396,31 +396,23 @@ void TransactionMgr::_clean_stream_context() {
     for (const auto& id : ids) {
         auto ctx = _exec_env->stream_context_mgr()->get(id);
         if (ctx != nullptr) {
-            int64_t now = UnixSeconds();
-            // try lock fail means transaction in processing
-            if (ctx->lock.try_lock()) {
-                // abort timeout transaction
-                if ((now - ctx->begin_txn_ts) > ctx->timeout_second && ctx->timeout_second > 0) {
-                    ctx->status = Status::Aborted(fmt::format("transaction is aborted by timeout."));
-                    auto st = _rollback_transaction(ctx);
-                    LOG(INFO) << "Abort transaction " << ctx->brief() << " since timeout " << ctx->timeout_second
-                              << " begin ts " << ctx->begin_txn_ts << " status " << st;
-                }
-
-                if (ctx->body_sink != nullptr) {
-                    if (!ctx->body_sink->exhausted()) {
-                        ctx->last_active_ts = UnixSeconds();
-                    }
-                }
-
-                if ((now - ctx->last_active_ts) > ctx->idle_timeout_sec + interval && ctx->idle_timeout_sec > 0) {
-                    ctx->status = Status::Aborted(fmt::format("transaction is aborted by idle timeout."));
+            Status status;
+            if (ctx->tsl_reach_timeout()) {
+                status = Status::Aborted(
+                        fmt::format("transaction is aborted by timeout {} seconds.", ctx->timeout_second));
+            } else if (ctx->tsl_reach_idle_timeout(interval)) {
+                status = Status::Aborted(
+                        fmt::format("transaction is aborted by idle timeout {} seconds.", ctx->idle_timeout_sec));
+            }
+            if (!status.ok()) {
+                ctx->timeout_detected.store(true, std::memory_order_release);
```

Comment:
shall this should be put into the lock? otherwise, the timeout_detected set to true and try_lock fails which causes inconsistent internal state.

Reply:
fixed